src/hive/tracer/zmq.clj

```diff
@@ -1,52 +1,51 @@
-            [clojure.core.async :as async]))
+  (:require [cheshire.core :as cheshire]
+            [clj-time.core :as t]
+            [clojure.core.async :as async]
+            [hive.tracer.adapters :as adapters]
+            [hive.config :as config]
+            [hive.tracer.heartbeat :as heartbeat]
+            [hive.storage.store :as store]
+            [zeromq.zmq :as zmq]))
 
-(def context (zmq/context 1))
-
-(defn bytes->string [^bytes bs]
-  (when bs
-    (String. bs)))
-
-(defn port->endpoint [port]
-  (str "tcp://*:" port))
+(def context (zmq/context config/io-threads))
 
 (defn new-router-socket! [port]
   (let [router (zmq/socket context :router)]
-    (zmq/set-receive-timeout router 1000)
-    (zmq/set-send-timeout router 1000)
-    (zmq/bind router (port->endpoint port))))
+    (zmq/set-receive-timeout router config/socket-receive-timeout)
+    (zmq/set-send-timeout router config/socket-send-timeout)
+    (zmq/bind router (adapters/port->endpoint port))))
 
 (defn receive-all-str! [socket]
-  (mapv bytes->string (zmq/receive-all socket)))
+  (mapv adapters/bytes->string (zmq/receive-all socket)))
 
-(defn receive-message! [socket]
+(defn try-receive-message! [socket]
   (let [[ident meta payload :as received] (receive-all-str! socket)]
     (when (every? some? received)
-      {:identity ident
-       :meta     (cheshire/parse-string meta true)
-       :payload  (cheshire/parse-string payload true)})))
+      (adapters/raw-event->internal received))))
 
 (defn respond! [router message]
-  (go (zmq/send router message)))
+  (async/go (zmq/send router message)))
 
 (defn start-receiving! [router on-receive]
-  (let [stop-channel (async/chan)
-        ch (async/chan 1000)]
+  (let [stop-ch      (async/chan)
+        heartbeat-ch (heartbeat/heartbeat-ch config/healthcheck-timing-s)
+        main-ch      (async/chan config/main-ch-buffer-size)]
     (async/go-loop []
-      (when (async/alt! stop-channel false :default :keep-going)
-        (some->> (receive-message! router)
-                 (async/>! ch))
+      (when (async/alt! stop-ch false :priority true :default :keep-going)
+        (some->> (try-receive-message! router)
+                 (async/>! main-ch))
         (recur)))
     (async/go-loop []
-      (when (async/alt! stop-channel false :default :keep-going)
-        (some-> (async/<! ch)
-                (on-receive router))
-        (recur)))
-    stop-channel))
+      (let [[source value] (async/alt! stop-ch [:stop]
+                                       main-ch ([v] [:main v])
+                                       heartbeat-ch ([v] [:heartbeat v]))]
+        (case source
+          :stop      (run! async/close! [stop-ch heartbeat-ch main-ch])
+          :main      (do (prn "Main channel triggered.") (on-receive value router) (recur))
+          :heartbeat (do (prn "Heartbeat check triggered.") (heartbeat/healthcheck-services!) (recur))
+          (recur))))
+    stop-ch))
```

Comment:
All this logic related to heartbeats should be in the heartbeat handler in `handlers.clj`

Reply:
What do you mean? this is the heartbeat trigger, not the logic.

Reply:
Two things should happen:
- When we receive a heatbeat the heartbeat handler should update the timestamp on the service-map
- We should start a loop to check this timestamp

None of they should happen in this function. The latter, should have it's own function and the started in the component startup (on `tracer/server.clj)`

Reply:
Yep they do. I think we should make everything just one self-contained component. It will be just a periodic loop checking it's internal state

Reply:
But I think I got your point, maybe it's an option. I was considering that we would move the atom to the tracer module. But it maybe a standalone component
I think we can do it

Reply:
3 components:
- ZMQServer
- PedestalServer
- Storage

Storage would have this loop checking it's state

Reply:
WDYT?

Reply:
:thinking: 
I think it makes sense. 
But the heartbeat would be with the zmqserver of with the storage comp?

Reply:
In the end if we want to create one channel for each service to trigger heartbeats instead of one global I see the separated component making more sense. But we can always do that later :+1:

Reply:
This is not trigging heartbeats, just checking the last timestamp. In this case the the verification would be in the storage component